### DISCUSSION TITLE
Best way to implement an hostname based router?

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#934] [Q&A] [answered]
Hello,

I would like to have different routes depending on the Hostname of the request.

Something similar to the following pseudo-code:

```go
router.Use(middleware.CleanPath)
router.Use(middleware.SetHTTPContext)
router.Use(middleware.Auth(kernelService, sitesService, membersService))
router.Use(middleware.Timeout(60 * time.Second))
router.Use(middleware.Recoverer)

router.Map("mydomain.com", websiteRoutes)
router.Map("api.mydomain.com", apiRoutes)

router.Map("*", otherRoutes)
```

Where `websiteRoutes`, `apiRoutes` and `otherRoutes` and standalone routers.

What is the best way to achieve that with axum?

Thank you 

## Edit

For those interested, I've published the complete code on my blog: https://kerkour.com/rust-axum-hostname-router

**Comment by davidpdrsn** [comment, marked as answer]
axum doesn't support routing directly based on the hostname. Instead you have to use the `Host` extractor, check the value in a handler, and forward the request to some other router.

**Reply by ghost** [reply]
Thank you very much!

I've got the following working:

```rust
    let api_router = Router::new().route("/*path", any(api_handler));
    let other_router = Router::new().route("/*path", any(other_handler));

    let app = Router::new().route("/*path", any(|Host(hostname): Host, request: Request<Body>| async move {
        match hostname.as_str() {
            "api.mydomain.com" =>  api_router.oneshot(request).await,
            _ => other_router.oneshot(request).await,
        }
    }));
```

Is this the right way to do it, or is there a "cleaner" way?


## Edit

For those interested, I've published the complete code on my blog: https://kerkour.com/rust-axum-hostname-router

**Reply by davidpdrsn** [reply]
That's fine

**Reply by ghost** [reply]
Great, thank you very much!

**Comment by axxop** [comment]
In axum6, we can use `middleware` to do this. 

```rust

let router = Router::new()
        .route("/", get(main))
       .with_state(())
      .layer(middleware::from_fn(my_middleware));

    async fn my_middleware<B>(Host(host): Host, request: Request<B>, next: Next<B>) -> Response {
        let lowcase_host = host.to_lowercase();
        if lowcase_host == "example.com" || lowcase_host == "www.example.com" {
            return next.run(request).await;
        }

        StatusCode::NOT_FOUND.into_response()
    }

```

Hope that helps others.
